plenum/common/stashing_router.py

```diff
@@ -0,0 +1,53 @@
+from collections import defaultdict
+from functools import partial
+from typing import Callable, Any, Dict, List, Type, Optional, Tuple
+
+DISCARD = -1
+PROCESS = 0
+STASH = 1
+
+
+class StashingRouter:
+    Handler = Callable[..., Optional[int]]
+
+    def __init__(self):
+        self._handlers = {}  # type: Dict[Type, StashingRouter.Handler]
+        self._queues = defaultdict(list)  # type: Dict[int, List[Tuple]]
+
+    def subscribe(self, message_type: Type, handler: Handler):
+        # TODO: Check double subscription?
+        self._handlers[message_type] = handler
+
+    def subscribe_to(self, bus: Any):
+        for message_type, handler in self._handlers.items():
+            bus.subscribe(message_type, partial(self._process, handler))
+
+    def unstash(self, code: Optional[int] = None):
+        if code is None:
+            for c in list(self._queues.keys()):
+                self._unstash(c)
+        else:
+            self._unstash(c)
+
+    def _process(self, handler: Handler, message: Any, *args):
+        result = handler(message, *args)
+
+        # If handler returned either None or PROCESS we assume it successfully processed message
+        # and no further action is needed
+        if not result:
+            return
+
+        if result == DISCARD:
+            # TODO: Log discard?
+            return
+
+        # TODO: Log stash?
+        self._queues[result].append((message, *args))
+
+    def _unstash(self, reason: int):
+        queue = self._queues[reason]
+        self._queues[reason] = []
+        for msg_tuple in queue:
+            msg_type = type(msg_tuple[0])
```

Comment:
Why `msg` is always a tuple? Should we mention it in `_process` then?

Reply:
Because depending on use case message handler can accept arbitrary number of arguments, first of which is always message. This is already reflected in signature of `_process`:
```python
    def _process(self, handler: Handler, message: Any, *args):
```
I would also like to reflect in Handler type annotation, doing something like
```python
    Handler = Callable[[Any, ...], Optional[int]]
```
however this is unfortunately an invalid syntax